Wrappers/Python/ccpi/optimisation/algorithms/CGLS.py

```diff
@@ -34,98 +34,145 @@ class CGLS(Algorithm):
       x_init: initial guess
       operator: operator for forward/backward projections
       data: data to operate on
+      tolerance: tolerance to stop algorithm
+      
+    Reference:
+        https://web.stanford.edu/group/SOL/software/cgls/
+      
     '''
+    
+    
+    
     def __init__(self, **kwargs):
+        
         super(CGLS, self).__init__()
         self.x        = kwargs.get('x_init', None)
         self.operator = kwargs.get('operator', None)
         self.data     = kwargs.get('data', None)
+        self.tolerance     = kwargs.get('tolerance', None)
         if self.x is not None and self.operator is not None and \
            self.data is not None:
-            print ("Calling from creator")
             self.set_up(x_init  =kwargs['x_init'],
                                operator=kwargs['operator'],
                                data    =kwargs['data'])
-
+            
+        if self.tolerance is None:
+            self.tolerance = 1e-6            
+                                    
     def set_up(self, x_init, operator , data ):
 
-        self.r = data.copy()
-        self.x = x_init * 0
-
-        self.operator = operator
-        self.d = operator.adjoint(self.r)
-
+        self.x = x_init
```

Comment:
Do we not need a `.copy()`?  I think without, once we update `x` we will also update `x_init` which is probably not what we want, for example if we then afterwards use `x_init` in another algorithm, it will be the final solution `x` instead of the zero image.

Reply:
well spotted

Reply:
well actually it needs `self.x = x_init * 0.`